File: 2024/Python/04.py
data = open("../inputs/04.txt").read().splitlines()

dydx = [(y, x) for y in range(-1, 2) for x in range(-1, 2) if y or x]
diag = [(-1, -1), (-1, 1), (1, -1), (1, 1)]

H = len(data)
W = len(data[0])

search_cnt = 0
revised_search_cnt = 0

for y, row in enumerate(data):
    for x, col in enumerate(row):
        if 0 < y < H - 1 and 0 < x < W - 1:
            corners = [data[y + dx][x + dy] for dx, dy in diag]
            if (
                col == "A"
                and corners.count("M") == 2
                and corners.count("S") == 2
                and corners[0] != corners[3]
            ):
                revised_search_cnt += 1

        if col != "X":
            continue

        for dy, dx in dydx:
            chars = ["X"]
            for d in range(1, 4):
                if 0 <= y + dy * d < H and 0 <= x + dx * d < W:
                    chars.append(data[y + dy * d][x + dx * d])
            if "".join(chars) == "XMAS":
                search_cnt += 1

print(search_cnt)
print(revised_search_cnt)

File: 2024/Deno/04.ts
const data = (await Deno.readTextFile("../inputs/04.txt")).trim().split("\n");

const dydx = [
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
];

const diag = [[-1, -1], [-1, 1], [1, -1], [1, 1]];

const H = data.length;
const W = data[0].length;

let searchCount = 0;
let revisedSearchCount = 0;

for (let y = 0; y < H; y++) {
    for (let x = 0; x < W; x++) {
        if (y > 0 && y < H - 1 && x > 0 && x < W - 1 && data[y][x] === "A") {
            const corners = diag.map(([dy, dx]) => data[y + dy][x + dx]);
            if (
                corners.filter((c) => c === "M").length === 2 &&
                corners.filter((c) => c === "S").length === 2 &&
                corners[0] !== corners[3]
            ) {
                revisedSearchCount++;
            }
        }

        if (data[y][x] != "X") continue;

        for (const [dy, dx] of dydx) {
            const chars = ["X"];

            for (let d = 1; d < 4; d++) {
                const newY = y + dy * d;
                const newX = x + dx * d;
                if (!(0 <= newY && newY < H && 0 <= newX && newX < W)) {
                    break;
                }
                chars.push(data[newY][newX]);
            }
            if (chars.join("") === "XMAS") {
                searchCount++;
            }
        }
    }
}

console.log(searchCount);
console.log(revisedSearchCount);
